File: year2017/day05/python/solution.py
def main():
  print(part_1())
  print(part_2())


def part_1():
  return simulate(input[:], lambda offset: 1)


def part_2():
  return simulate(input[:], lambda offset: -1 if offset >= 3 else 1)


def setup(path):
  global input

  with open(path, 'r') as file:
    input = [int(num) for num in file.read().split('\n')]


def simulate(inst, offset_rule):
  ptr, jumps = 0, 0
  while 0 <= ptr < len(inst):
    inst[ptr], ptr, jumps = inst[ptr] + offset_rule(inst[ptr]), ptr + inst[ptr], jumps + 1

  return jumps


if __name__ == '__main__':
  setup('../puzzle.txt')
  main()

File: year2017/day05/java/Solution.java
import java.nio.file.*;
import java.util.*;
import java.util.function.IntUnaryOperator;

public class Solution {
    private static List<Integer> NUMBERS = new ArrayList<Integer>();

    public static void main(String[] args) { 
        try {
            setup(Files.readString(Paths.get("../puzzle.txt")));
            System.out.println(part1());
            System.out.println(part2());
        } catch(Exception e) {
            e.printStackTrace();
        }
    }
    
	public static int part1() {
        return simulate(NUMBERS.stream().mapToInt(Integer::intValue).toArray(), offset -> 1);
    }
    
    public static int part2() {
        return simulate(NUMBERS.stream().mapToInt(Integer::intValue).toArray(), offset -> (offset >= 3 ? -1 : 1));
    }

    public static void setup(String input) {
        for(String num : input.split("\n")) {
            NUMBERS.add(Integer.parseInt(num));
        }
    }

    private static int simulate(int[] inst, IntUnaryOperator offsetRule) {
        int ptr = 0;
        int jumps = 0;

        while (ptr >= 0 && ptr < inst.length) {
            int offset = inst[ptr];
            inst[ptr] = inst[ptr] + offsetRule.applyAsInt(offset);
            ptr = ptr + offset;
            jumps++;
        }

        return jumps;
    }
}